src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -169,20 +173,50 @@ public async Task SendRequestBodyAsync(CancellationToken cancellationToken)
                 {
                     if (NetEventSource.IsEnabled) Trace($"Failed to send request body: {e}");
 
-                    // Cancel the stream before we set _requestCompletionState below.
-                    // Otherwise, a response stream reader may race with the actual Cancel.
-                    Cancel();
+                    bool signalWaiter = false;
 
                     lock (SyncObject)
                     {
                         Debug.Assert(_requestCompletionState == StreamCompletionState.InProgress, $"Request already completed with state={_requestCompletionState}");
 
-                        _requestCompletionState = StreamCompletionState.Failed;
+                        if (_requestBodyAbandoned)
+                        {
+                            // See comments on _requestBodyAbandoned.
+                            // In this case, the request is still considered successful and we do not want to send a RST_STREAM, 
+                            // and we also don't want to propagate any error to the caller, in particular for non-duplex scenarios.
+                            Debug.Assert(_responseCompletionState == StreamCompletionState.Completed);
+                            _requestCompletionState = StreamCompletionState.Completed;
+                            Complete();
+                            return;
+                        }
+                        else
+                        {
+                            _requestCompletionState = StreamCompletionState.Failed;
+
+                            // Cancel the response body, if not complete.
+                            if (_responseCompletionState == StreamCompletionState.InProgress)
+                            {
+                                _responseCompletionState = StreamCompletionState.Failed;
+                            }
+
+                            // Discard any remaining buffered response data
+                            if (_responseBuffer.ActiveLength != 0)
+                            {
+                                _responseBuffer.Discard(_responseBuffer.ActiveLength);
+                            }
 
-                        // Cancel above should ensure that the response is either Completed or Failed now.
-                        Debug.Assert(_responseCompletionState != StreamCompletionState.InProgress);
+                            _responseProtocolState = ResponseProtocolState.Aborted;
 
-                        Reset();
+                            signalWaiter = _hasWaiter;
+                            _hasWaiter = false;
+
+                            Reset();
```

Comment:
Why can't this else block use Cancel?  At a minimum please add a comment explaining. But it feels like we're duplicating logic here that it would be good to consolidate.

Reply:
It can't use Cancel as written because Cancel currently takes the lock.

I could of course refactor so that Cancel doesn't take the lock, but callers of it do instead. But then, Cancel also signals the waiter if necessary and cancels the cancellation token, and it does these after releasing the lock. I could of course deal with that too, but now we just start getting in to more and more code change, and I'm reluctant to refactor the code so much at this late stage. So I basically opted to just duplicate some code.

If you think I'm being too conservative here, let me know.

I'm happy to add a comment.